### Meaning of Columns
Department, Gender, HSC - Score obtained in Higher secondary education, SSC - Score obtained in secondary school education, Income - Family monthly income of parents, Hometown, Computer - proficiency level in computer usage, Preparation - TIme spent on study preparation outside class hours, Gaming - Time spent on gaming activities daily, Attendance, Job - if student has a part time job, English - Proficiency in English commmunication skills, Extra - Participation in extracurricular activities, Semester - Current semester the student is enrolled in, Last - Performance in the last semester, Overall - CGPA

This dataset can be used to predict if a student will finish school with a first class grade or not (or can be used to predict the final grading for students - grade A, B, C, D, E, F)

### Questions to Ask
1. Analyze the relationships between student demographics, academic achievements, socio-economic factors and extra-curricular activities - correlation
2. No of male and female above a certian CGPA
3. Monthly income range of parents of student within certain CGPA ranges
4. If their Hometown affects the CGPA and dedication to school activities
5. Impact of Attendance on CGPA
6. Impact of having a part-time job on attendance, CGPA, extracurrcular activities, preparation
7. Impact of computer proficiency and English proficiency on CGPA and last semester performance
8. Impact of Gaming time on preparation, attendance, last semester performance and CGPA
9. Impact of parent's financial income on students, the need to get a part time job, participation in extracurricular activities, last semester performance and CGPA
10. correlation between attendance and CGPA, between attendance and Gaming, between attendance and job, attendance and last semester performance.
11. Gender with the highest CGPA
12. Gender with the best performance last semester
13. Gender involved in extracurricular activities more
14. correlation between higher secondary education score and CGPA, last semester performance
15. correlation between secondary school education and CGPA, last semester performance

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# to ensure that the number of columns seen is extended
pd.set_option('display.max_columns', 100)

In [3]:
data = pd.read_csv(r'C:\Users\chose\alexanderchosen\Personal\ML\eda_with_python\student_performance_metrics\student_performance_metric.csv.csv')

### Understanding the dataset

In [4]:
data.columns

Index(['Department', 'Gender', 'HSC', 'SSC', 'Income', 'Hometown', 'Computer',
       'Preparation', 'Gaming', 'Attendance', 'Job', 'English', 'Extra',
       'Semester', 'Last', 'Overall'],
      dtype='object')

In [5]:
data.head(20)

Department  Gender   HSC   SSC  \
0            Business Administration    Male  4.17  4.84   
1            Business Administration  Female  4.92  5.00   
2            Business Administration    Male  5.00  4.83   
3            Business Administration    Male  4.00  4.50   
4            Business Administration  Female  2.19  3.17   
5   Computer Science and Engineering    Male  4.75  4.05   
6   Computer Science and Engineering    Male  4.42  5.00   
7   Computer Science and Engineering    Male  4.50  4.81   
8   Computer Science and Engineering    Male  3.32  4.50   
9   Computer Science and Engineering  Female  3.33  4.95   
10  Computer Science and Engineering    Male  4.17  5.00   
11  Computer Science and Engineering    Male  3.58  4.56   
12  Computer Science and Engineering    Male  4.50  4.75   
13  Computer Science and Engineering    Male  4.50  4.75   
14  Computer Science and Engineering    Male  3.83  5.00   
15  Computer Science and Engineering    Male  4.33  4.64   
16  Computer Science and Engineering  Female  4.51  4.75   
17  Computer Science and Engineering    Male  4.50  4.75   
18  Computer Science and Engineering    Male  3.92  4.72   
19  Computer Science and Engineering    Male  3.92  5.00   

                           Income Hometown  Computer        Preparation  \
0              Low (Below 15,000)  Village         3  More than 3 Hours   
1    Upper middle (30,000-50,000)     City         3           0-1 Hour   
2    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
3             High (Above 50,000)     City         5  More than 3 Hours   
4    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
5    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
6             High (Above 50,000)  Village         4           0-1 Hour   
7    Upper middle (30,000-50,000)     City         3          2-3 Hours   
8             Low (Below 15,000)      City         4           0-1 Hour   
9   Lower middle (15,000-30,000)      City         3           0-1 Hour   
10   Lower middle (15,000-30,000)  Village         3           0-1 Hour   
11   Upper middle (30,000-50,000)  Village         4           0-1 Hour   
12   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
13   Upper middle (30,000-50,000)  Village         1           0-1 Hour   
14             Low (Below 15,000)  Village         2           0-1 Hour   
15  Lower middle (15,000-30,000)      City         5           0-1 Hour   
16   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
17   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
18            High (Above 50,000)  Village         3           0-1 Hour   
19            High (Above 50,000)  Village         4           0-1 Hour   

               Gaming Attendance Job  English Extra Semester   Last  Overall  
0            0-1 Hour   80%-100%  No        3   Yes      6th  3.220    3.350  
1            0-1 Hour   80%-100%  No        3   Yes      7th  3.467    3.467  
2   More than 3 Hours   80%-100%  No        4   Yes      3rd  4.000    3.720  
3   More than 3 Hours   80%-100%  No        5   Yes      4th  3.800    3.750  
4           2-3 Hours   80%-100%  No        3   Yes      4th  3.940    3.940  
5   More than 3 Hours  Below 40%  No        4    No      2nd  1.000    1.000  
6   More than 3 Hours    60%-79%  No        2    No      2nd  1.060    1.060  
7   More than 3 Hours   80%-100%  No        4   Yes     11th  2.950    1.250  
8   More than 3 Hours  Below 40%  No        3   Yes      5th  1.420    1.440  
9   More than 3 Hours  Below 40%  No        4    No      2nd  1.500    1.500  
10  More than 3 Hours  Below 40%  No        4    No      2nd  1.560    1.560  
11  More than 3 Hours  Below 40%  No        4    No      9th  1.570    1.570  
12  More than 3 Hours    60%-79%  No        2   Yes      8th  1.570    1.570  
13  More than 3 Hours    60%-79%  No        2    No      8th  1.000    1.600  
14  

In [6]:
data.tail()

Department  Gender   HSC   SSC                        Income Hometown  \
488  Public Health    Male  3.50  4.00  Lower middle (15,000-30,000)  Village   
489  Public Health    Male  4.50  4.00  Upper middle (30,000-50,000)     City   
490      Sociology  Female  4.56  3.67  Lower middle (15,000-30,000)     City   
491      Sociology    Male  4.23  4.50            Low (Below 15,000)  Village   
492      Sociology  Female  3.83  5.00  Lower middle (15,000-30,000)  Village   

     Computer        Preparation     Gaming Attendance  Job  English Extra  \
488         2  More than 3 Hours  2-3 Hours    40%-59%  Yes        2   Yes   
489         4          2-3 Hours  2-3 Hours    60%-79%   No        3   Yes   
490         3           0-1 Hour  2-3 Hours    40%-59%   No        2    No   
491         3  More than 3 Hours  2-3 Hours   80%-100%   No        3   Yes   
492         3  More than 3 Hours  2-3 Hours   80%-100%   No        4   Yes   

    Semester  Last  Overall  
488      7th  3.25     3.10  
489      9th  3.50     3.67  
490      2nd  3.56     3.56  
491     10th  3.61     3.75  
492     12th  4.00     3.86

In [7]:
data.shape

(493, 16)

In [8]:
data.dtypes

Department      object
Gender          object
HSC            float64
SSC            float64
Income          object
Hometown        object
Computer         int64
Preparation     object
Gaming          object
Attendance      object
Job             object
English          int64
Extra           object
Semester        object
Last           float64
Overall        float64
dtype: object

In [9]:
data.describe()

HSC         SSC    Computer     English        Last     Overall
count  493.000000  493.000000  493.000000  493.000000  493.000000  493.000000
mean     4.156775    4.767647    3.338742    3.569980    3.163888    3.188383
std      0.546504    0.349538    1.252988    0.870515    0.641475    0.591801
min      2.170000    3.000000    1.000000    1.000000    1.000000    1.000000
25%      3.830000    4.680000    3.000000    3.000000    2.810000    2.880000
50%      4.170000    4.940000    3.000000    4.000000    3.250000    3.270000
75%      4.500000    5.000000    4.000000    4.000000    3.670000    3.680000
max      5.000000    5.000000    5.000000    5.000000    4.000000    4.000000

In [10]:
data['Computer'].value_counts()

Computer
3    183
5    115
4     95
1     58
2     42
Name: count, dtype: int64

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 493 entries, 0 to 492
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Department   493 non-null    object 
 1   Gender       493 non-null    object 
 2   HSC          493 non-null    float64
 3   SSC          493 non-null    float64
 4   Income       493 non-null    object 
 5   Hometown     493 non-null    object 
 6   Computer     493 non-null    int64  
 7   Preparation  493 non-null    object 
 8   Gaming       493 non-null    object 
 9   Attendance   493 non-null    object 
 10  Job          493 non-null    object 
 11  English      493 non-null    int64  
 12  Extra        493 non-null    object 
 13  Semester     493 non-null    object 
 14  Last         493 non-null    float64
 15  Overall      493 non-null    float64
dtypes: float64(4), int64(2), object(10)
memory usage: 61.8+ KB


In [12]:
data.shape

(493, 16)

This indicates that the dataset has 493 rows and 16 columns. It has no null values, and all the columns are needed for this analysis. However, I need to convert some of the columns to numerical data or a sort of grade. Examples are the Attendance, preparation and gaming

In [13]:
# to convert preparation column to numeric grades
# that is, 1 for 0-1 hour, 2 for 2-3 hours, 3 for more than 3 hours
data['Preparation'].value_counts()

Preparation
2-3 Hours            228
0-1 Hour             218
More than 3 Hours     47
Name: count, dtype: int64

In [15]:
data['Preparation'].unique()

array(['More than 3 Hours', '0-1 Hour', '2-3 Hours'], dtype=object)

In [22]:
# define mapping using d ictionary
preparation_dict = {
    "0-1 Hour": 1,
    "2-3 Hours": 2,
    "More than 3 Hours": 3
}

preparation_dict


{'0-1 Hour': 1, '2-3 Hours': 2, 'More than 3 Hours': 3}

In [23]:
# use the dictionary to create a new column
data['preparation_level'] = data['Preparation'].map(preparation_dict)
data.head(20)

Department  Gender   HSC   SSC  \
0            Business Administration    Male  4.17  4.84   
1            Business Administration  Female  4.92  5.00   
2            Business Administration    Male  5.00  4.83   
3            Business Administration    Male  4.00  4.50   
4            Business Administration  Female  2.19  3.17   
5   Computer Science and Engineering    Male  4.75  4.05   
6   Computer Science and Engineering    Male  4.42  5.00   
7   Computer Science and Engineering    Male  4.50  4.81   
8   Computer Science and Engineering    Male  3.32  4.50   
9   Computer Science and Engineering  Female  3.33  4.95   
10  Computer Science and Engineering    Male  4.17  5.00   
11  Computer Science and Engineering    Male  3.58  4.56   
12  Computer Science and Engineering    Male  4.50  4.75   
13  Computer Science and Engineering    Male  4.50  4.75   
14  Computer Science and Engineering    Male  3.83  5.00   
15  Computer Science and Engineering    Male  4.33  4.64   
16  Computer Science and Engineering  Female  4.51  4.75   
17  Computer Science and Engineering    Male  4.50  4.75   
18  Computer Science and Engineering    Male  3.92  4.72   
19  Computer Science and Engineering    Male  3.92  5.00   

                           Income Hometown  Computer        Preparation  \
0              Low (Below 15,000)  Village         3  More than 3 Hours   
1    Upper middle (30,000-50,000)     City         3           0-1 Hour   
2    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
3             High (Above 50,000)     City         5  More than 3 Hours   
4    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
5    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
6             High (Above 50,000)  Village         4           0-1 Hour   
7    Upper middle (30,000-50,000)     City         3          2-3 Hours   
8             Low (Below 15,000)      City         4           0-1 Hour   
9   Lower middle (15,000-30,000)      City         3           0-1 Hour   
10   Lower middle (15,000-30,000)  Village         3           0-1 Hour   
11   Upper middle (30,000-50,000)  Village         4           0-1 Hour   
12   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
13   Upper middle (30,000-50,000)  Village         1           0-1 Hour   
14             Low (Below 15,000)  Village         2           0-1 Hour   
15  Lower middle (15,000-30,000)      City         5           0-1 Hour   
16   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
17   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
18            High (Above 50,000)  Village         3           0-1 Hour   
19            High (Above 50,000)  Village         4           0-1 Hour   

               Gaming Attendance Job  English Extra Semester   Last  Overall  \
0            0-1 Hour   80%-100%  No        3   Yes      6th  3.220    3.350   
1            0-1 Hour   80%-100%  No        3   Yes      7th  3.467    3.467   
2   More than 3 Hours   80%-100%  No        4   Yes      3rd  4.000    3.720   
3   More than 3 Hours   80%-100%  No        5   Yes      4th  3.800    3.750   
4           2-3 Hours   80%-100%  No        3   Yes      4th  3.940    3.940   
5   More than 3 Hours  Below 40%  No        4    No      2nd  1.000    1.000   
6   More than 3 Hours    60%-79%  No        2    No      2nd  1.060    1.060   
7   More than 3 Hours   80%-100%  No        4   Yes     11th  2.950    1.250   
8   More than 3 Hours  Below 40%  No        3   Yes      5th  1.420    1.440   
9   More than 3 Hours  Below 40%  No        4    No      2nd  1.500    1.500   
10  More than 3 Hours  Below 40%  No        4    No      2nd  1.560    1.560   
11  More than 3 Hours  Below 40%  No        4    No      9th  1.570    1.570   
12  More than 3 Hours    60%-79%  No        2   Yes      8th  1.570    1.570   
13  More than 3 Hours    60%-79%  No        2    No      8th  1.000  

In [24]:
data[['Preparation', 'preparation_level']].head()

Preparation  preparation_level
0  More than 3 Hours                  3
1           0-1 Hour                  1
2           0-1 Hour                  1
3  More than 3 Hours                  3
4           0-1 Hour                  1

In [27]:
data['preparation_level'].value_counts()

preparation_level
2    228
1    218
3     47
Name: count, dtype: int64

In [28]:
# to convert the gaming column data from a set of range values to a numeric value representing each range
data['Gaming'].unique()

array(['0-1 Hour', 'More than 3 Hours', '2-3 Hours'], dtype=object)

In [29]:
# next, after checking its unique values, create a gaming dictionary
gaming_dict = {
    "0-1 Hour":1,
    "2-3 Hours": 2,
    "More than 3 Hours": 3
} 

gaming_dict

{'0-1 Hour': 1, '2-3 Hours': 2, 'More than 3 Hours': 3}

In [31]:
# next, map the dictionary values to the new column
data['gaming_level'] = data['Gaming'].map(gaming_dict)
data.head(20)

Department  Gender   HSC   SSC  \
0            Business Administration    Male  4.17  4.84   
1            Business Administration  Female  4.92  5.00   
2            Business Administration    Male  5.00  4.83   
3            Business Administration    Male  4.00  4.50   
4            Business Administration  Female  2.19  3.17   
5   Computer Science and Engineering    Male  4.75  4.05   
6   Computer Science and Engineering    Male  4.42  5.00   
7   Computer Science and Engineering    Male  4.50  4.81   
8   Computer Science and Engineering    Male  3.32  4.50   
9   Computer Science and Engineering  Female  3.33  4.95   
10  Computer Science and Engineering    Male  4.17  5.00   
11  Computer Science and Engineering    Male  3.58  4.56   
12  Computer Science and Engineering    Male  4.50  4.75   
13  Computer Science and Engineering    Male  4.50  4.75   
14  Computer Science and Engineering    Male  3.83  5.00   
15  Computer Science and Engineering    Male  4.33  4.64   
16  Computer Science and Engineering  Female  4.51  4.75   
17  Computer Science and Engineering    Male  4.50  4.75   
18  Computer Science and Engineering    Male  3.92  4.72   
19  Computer Science and Engineering    Male  3.92  5.00   

                           Income Hometown  Computer        Preparation  \
0              Low (Below 15,000)  Village         3  More than 3 Hours   
1    Upper middle (30,000-50,000)     City         3           0-1 Hour   
2    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
3             High (Above 50,000)     City         5  More than 3 Hours   
4    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
5    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
6             High (Above 50,000)  Village         4           0-1 Hour   
7    Upper middle (30,000-50,000)     City         3          2-3 Hours   
8             Low (Below 15,000)      City         4           0-1 Hour   
9   Lower middle (15,000-30,000)      City         3           0-1 Hour   
10   Lower middle (15,000-30,000)  Village         3           0-1 Hour   
11   Upper middle (30,000-50,000)  Village         4           0-1 Hour   
12   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
13   Upper middle (30,000-50,000)  Village         1           0-1 Hour   
14             Low (Below 15,000)  Village         2           0-1 Hour   
15  Lower middle (15,000-30,000)      City         5           0-1 Hour   
16   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
17   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
18            High (Above 50,000)  Village         3           0-1 Hour   
19            High (Above 50,000)  Village         4           0-1 Hour   

               Gaming Attendance Job  English Extra Semester   Last  Overall  \
0            0-1 Hour   80%-100%  No        3   Yes      6th  3.220    3.350   
1            0-1 Hour   80%-100%  No        3   Yes      7th  3.467    3.467   
2   More than 3 Hours   80%-100%  No        4   Yes      3rd  4.000    3.720   
3   More than 3 Hours   80%-100%  No        5   Yes      4th  3.800    3.750   
4           2-3 Hours   80%-100%  No        3   Yes      4th  3.940    3.940   
5   More than 3 Hours  Below 40%  No        4    No      2nd  1.000    1.000   
6   More than 3 Hours    60%-79%  No        2    No      2nd  1.060    1.060   
7   More than 3 Hours   80%-100%  No        4   Yes     11th  2.950    1.250   
8   More than 3 Hours  Below 40%  No        3   Yes      5th  1.420    1.440   
9   More than 3 Hours  Below 40%  No        4    No      2nd  1.500    1.500   
10  More than 3 Hours  Below 40%  No        4    No      2nd  1.560    1.560   
11  More than 3 Hours  Below 40%  No        4    No      9th  1.570    1.570   
12  More than 3 Hours    60%-79%  No        2   Yes      8th  1.570    1.570   
13  More than 3 Hours    60%-79%  No        2    No      8th  1.000  

In [32]:
# Finally, to convert the Attendance column into Attendance levels
# we check for the unique values first
data['Attendance'].unique()

array(['80%-100%', 'Below 40%', '60%-79%', '40%-59%'], dtype=object)

In [35]:
# next, i will create a dictionary
attendance_dict ={
    "80%-100%": 4,
    "60%-79%": 3,
    "40%-59%": 2,
    "Below 40%": 1
}
attendance_dict

{'80%-100%': 4, '60%-79%': 3, '40%-59%': 2, 'Below 40%': 1}

In [36]:
# finally, map the dictionary to the new column 
data['attendance_level']= data['Attendance'].map(attendance_dict)
data.head(20)

Department  Gender   HSC   SSC  \
0            Business Administration    Male  4.17  4.84   
1            Business Administration  Female  4.92  5.00   
2            Business Administration    Male  5.00  4.83   
3            Business Administration    Male  4.00  4.50   
4            Business Administration  Female  2.19  3.17   
5   Computer Science and Engineering    Male  4.75  4.05   
6   Computer Science and Engineering    Male  4.42  5.00   
7   Computer Science and Engineering    Male  4.50  4.81   
8   Computer Science and Engineering    Male  3.32  4.50   
9   Computer Science and Engineering  Female  3.33  4.95   
10  Computer Science and Engineering    Male  4.17  5.00   
11  Computer Science and Engineering    Male  3.58  4.56   
12  Computer Science and Engineering    Male  4.50  4.75   
13  Computer Science and Engineering    Male  4.50  4.75   
14  Computer Science and Engineering    Male  3.83  5.00   
15  Computer Science and Engineering    Male  4.33  4.64   
16  Computer Science and Engineering  Female  4.51  4.75   
17  Computer Science and Engineering    Male  4.50  4.75   
18  Computer Science and Engineering    Male  3.92  4.72   
19  Computer Science and Engineering    Male  3.92  5.00   

                           Income Hometown  Computer        Preparation  \
0              Low (Below 15,000)  Village         3  More than 3 Hours   
1    Upper middle (30,000-50,000)     City         3           0-1 Hour   
2    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
3             High (Above 50,000)     City         5  More than 3 Hours   
4    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
5    Lower middle (15,000-30,000)  Village         3           0-1 Hour   
6             High (Above 50,000)  Village         4           0-1 Hour   
7    Upper middle (30,000-50,000)     City         3          2-3 Hours   
8             Low (Below 15,000)      City         4           0-1 Hour   
9   Lower middle (15,000-30,000)      City         3           0-1 Hour   
10   Lower middle (15,000-30,000)  Village         3           0-1 Hour   
11   Upper middle (30,000-50,000)  Village         4           0-1 Hour   
12   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
13   Upper middle (30,000-50,000)  Village         1           0-1 Hour   
14             Low (Below 15,000)  Village         2           0-1 Hour   
15  Lower middle (15,000-30,000)      City         5           0-1 Hour   
16   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
17   Lower middle (15,000-30,000)  Village         1           0-1 Hour   
18            High (Above 50,000)  Village         3           0-1 Hour   
19            High (Above 50,000)  Village         4           0-1 Hour   

               Gaming Attendance Job  English Extra Semester   Last  Overall  \
0            0-1 Hour   80%-100%  No        3   Yes      6th  3.220    3.350   
1            0-1 Hour   80%-100%  No        3   Yes      7th  3.467    3.467   
2   More than 3 Hours   80%-100%  No        4   Yes      3rd  4.000    3.720   
3   More than 3 Hours   80%-100%  No        5   Yes      4th  3.800    3.750   
4           2-3 Hours   80%-100%  No        3   Yes      4th  3.940    3.940   
5   More than 3 Hours  Below 40%  No        4    No      2nd  1.000    1.000   
6   More than 3 Hours    60%-79%  No        2    No      2nd  1.060    1.060   
7   More than 3 Hours   80%-100%  No        4   Yes     11th  2.950    1.250   
8   More than 3 Hours  Below 40%  No        3   Yes      5th  1.420    1.440   
9   More than 3 Hours  Below 40%  No        4    No      2nd  1.500    1.500   
10  More than 3 Hours  Below 40%  No        4    No      2nd  1.560    1.560   
11  More than 3 Hours  Below 40%  No        4    No      9th  1.570    1.570   
12  More than 3 Hours    60%-79%  No        2   Yes      8th  1.570    1.570   
13  More than 3 Hours    60%-79%  No        2    No      8th  1.000  